### PR TITLE
Histogram logging feature

### PR DESCRIPTION
**Pull request by cronburg**
This PR / feature adds completion latency histogram output to fio, piggybacking on the existing histograms recorded by stat.c and adding the following command line options:

- `--log_hist_msec` operates like `--log_avg_msec` does, but produces files
  named `*_clat_hist*`
- `--write_hist_log` has the same semantics as `--write_lat_log` but for histograms.
- `--log_hist_coarseness` changes how many bins are logged.

The more substantial contribution is a post-processor for digesting these log files, located in `fio/tools/hist/fiologparser_hist.py`, which computes min/max/avg/percentile statistics.

To compare the accuracy of the statistics produced by `fiologparser_hist.py`, I also have a [reference implementation in my histograms branch of my fio fork]( https://github.com/cronburg/fio/blob/histograms/tools/hist/fiologparser_numpy.py) for computing the same statistics from the `*_clat.*` files fio already produces. The comparison can be viewed here: [D3 Latency Plotting Demo](https://cronburg.com/fio/demo/latency.html)

I'm excluding the reference implementation from this PR, as I'm still working with Mark on getting it into the branch where he's working on `fiologparser`. Also I'd eventually like to make `FIO_IO_U_PLAT_GROUP_NR` in `stat.h` configurable (allowing for a larger maximum histogram latency value than 17 seconds), but haven't added it to keep the changes to fio minimal for now.

A more detailed description of the motivation behind this feature, and explanation of the design process and algorithms used in the post-processor can be found here: [Cloud Latency Measurement](https://cronburg.com/fio/cloud-latency-problem-measurement/)

While I will not be available 24/7 to support this feature, the Red Hat performance team is planning on using this functionality to investigate effect of various configuration change events in a Ceph cluster on application latency, in order to improve regression testing and improve testing of hardware failure conditions, which are more common in scale-out storage configurations.

### CODE REVIEW

File: stat.c [hunk 1]
```diff
@@ -2205,6 +2208,35 @@ void add_clat_sample(struct thread_data *td, enum fio_ddir ddir,
...
+		
+		if (this_window >= iolog->hist_msec) {
+			/*
+			 * Make a byte-for-byte copy of the latency histogram stored in
+			 * td->ts.io_u_plat[ddir], recording it in a log sample. Note that
+			 * the matching call to sfree() is located in iolog.c after printing
+			 * this sample to the log file.
+			 */
+			unsigned int *io_u_plat = (unsigned int*)(td->ts.io_u_plat[ddir]);
+			unsigned int* dst = smalloc(FIO_IO_U_PLAT_NR * sizeof(unsigned int));
```
Comment (axboe) [inline stat.c]: Does this need to be smalloc'ed memory? It's a precious resource.

**Comment by axboe** [comment]
Looks good in general! A few things I'd like you go correct:

1) You have a great description here, yet the commit message is... non-existent. Basically everything in here should be in the commit message.
2) You are missing fio.1 and HOWTO update.
3) There are a few style issues here and there, like a space between '*' and the variable, etc.

File: options.c [hunk 2]
```diff
@@ -3530,6 +3530,37 @@ struct fio_option fio_options[FIO_MAX_OPTS] = {
...
+		.type = FIO_OPT_INT,
+		.off1 = td_var_offset(log_hist_msec),
+		.help = "Dump bw/iops/lat histograms at frequency of this time value",
+		.def  = "0",
+		.category = FIO_OPT_C_LOG,
+		.group = FIO_OPT_G_INVALID,
+	},
+	{
+		.name = "log_hist_coarseness",
+		.lname = "Log histograms (msec)",
```
Comment (axboe) [inline options.c]: Wrong long name?

**Comment by cronburg** [comment]
Are links in the commit message fine or should I not reference those?

I added / fixed:
- Commit message
- fio.1 and HOWTO information
- one instance of space between '*' and variable - will look through again to see if I missed any
- malloc / free instead of smalloc / sfree
- long name copy and paste error in options.c

**Comment by axboe** [comment]
Links in the commit message are fine. On the style issue, another thing I noticed is when you break up a line. The comma after a variable should be on the previous line, not the first thing on a new one.

**Comment by cronburg** [comment]
Fixed the commas, and some tabbing in options.c not matching the style around it.

**Comment by axboe** [comment]
Close enough! Thanks for doing this work. I committed two minor fixups:

1) Server version protocol update, along with transferring and converting the log file name appropriately
2) Minor coding style fixups

**Pull request merged**